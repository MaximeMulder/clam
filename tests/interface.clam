type IntInterface = [T] {
    zero: T,
    prec: (T) -> T,
    succ: (T) -> T,
    is_zero: (T) -> Bool,
    print: (T) -> Void
};

type MyInt = Int;

def my_int: IntInterface[MyInt] = @{
    zero = 0,
    prec = (i) => i - 1,
    succ = (i) => i + 1,
    is_zero = (i) => i == 0,
    print = print
};

def apply_n = [T] => (v: T, f: (T) -> T, n: Int) -> T =>
    if n == 0 then
        v
    else
        apply_n[T](f(v), f, n - 1);

def add_10 = [T] => (t: IntInterface[T], i: T) =>
    apply_n[T](i, t.succ, 10);

def main = {
    var zero = my_int.zero;
    my_int.print(zero);
    print(my_int.is_zero(zero));
    var ten = add_10[MyInt](my_int, zero);
    my_int.print(ten);
    print(my_int.is_zero(ten));
};
