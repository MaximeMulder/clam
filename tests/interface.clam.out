0
true
10
false
